tsdb/head.go

```diff
@@ -1622,15 +1624,57 @@ func (h *headIndexReader) SortedLabelValues(name string) ([]string, error) {
 
 // LabelValues returns label values present in the head for the
 // specific label name that are within the time range mint to maxt.
-func (h *headIndexReader) LabelValues(name string) ([]string, error) {
+// If matchers are specified the returned result set is reduced
+// to label values of metrics matching the matchers.
+func (h *headIndexReader) LabelValues(name string, matchers ...*labels.Matcher) ([]string, error) {
 	h.head.symMtx.RLock()
-	defer h.head.symMtx.RUnlock()
+
 	if h.maxt < h.head.MinTime() || h.mint > h.head.MaxTime() {
+		h.head.symMtx.RUnlock()
 		return []string{}, nil
 	}
 
-	values := h.head.postings.LabelValues(name)
-	return values, nil
+	if len(matchers) == 0 {
+		h.head.symMtx.RUnlock()
+		return h.head.postings.LabelValues(name), nil
+	}
+
+	// We're only interested in metrics which have the label <name>.
+	requireLabel, err := labels.NewMatcher(labels.MatchNotEqual, name, "")
+	if err != nil {
+		h.head.symMtx.RUnlock()
+		return nil, errors.Wrapf(err, "Failed to instantiate label matcher")
+	}
+
+	postings, err := PostingsForMatchers(h, append(matchers, requireLabel)...)
+	if err != nil {
+		h.head.symMtx.RUnlock()
+		return nil, err
+	}
+
+	seenResults := make(map[string]interface{})
+	for postings.Next() {
+		memSeries := h.head.series.getByID(postings.At())
+		if memSeries == nil {
+			continue
+		}
+
+		seenResults[memSeries.lset.Get(name)] = nil
```

Comment:
Based on what I commented about the `symMtx`, we can use the `LabelValueFor` here and maybe even share the code between head.go and block.go to do this.

Reply:
That's a very nice idea, how about that: d91d7f2d93eaa25a085627d618722073025b9b50

I didn't really know where to put this `labelValuesWithMatchers()` function because its supposed to be generic and not specific to either the head or the block, so i added a `util.go` file... Do you have a better idea?

Reply:
Since `PostingsForMatchers` is in querier.go and as this is related to queries, lets get rid of util.go and put it in querier.go?

Reply:
good idea, done https://github.com/prometheus/prometheus/pull/8400/commits/3c3b76a02a95f6aa64d1d100f28793fa63ad9292